app/controllers/PayPalOneOff.scala

```diff
@@ -54,21 +57,30 @@ class PayPalOneOff(
       "payerId" -> PayerID
     )
 
-    def processPaymentApiResponse(success: Boolean): Result = {
-      if (success) {
-        SafeLogger.info(s"One-off contribution for Paypal payment is successful. Sending user to thank-you page")
-        Redirect("/contribute/one-off/thankyou")
-      } else {
-        SafeLogger.error(scrub"Error making paypal payment. Sending user to error page.")
-        Ok(views.html.main(
-          "Support the Guardian | PayPal Error",
-          "paypal-error-page",
-          "payPalErrorPage.js",
-          "payPalErrorPageStyles.css"
-        ))
-      }
-    }
-
+    val paypalErrorPage = Ok(views.html.main(
+      "Support the Guardian | PayPal Error",
+      "paypal-error-page",
+      "payPalErrorPage.js",
+      "payPalErrorPageStyles.css"
+    ))
+    def processPaymentApiResponse(paymentApiResponse: Option[PaymentAPIResponse[PayPalError, PaypalExecuteSuccessResponse]]): Result =
+      paymentApiResponse.fold(paypalErrorPage)(resp => {
+        resp.data.fold(
```

Comment:
I find it a bit difficult to follow all these folds. Could you use pattern matching to make it easier to read?

```    
      paymentApiResponse match {
        case Some(PaymentAPIResponse(_, None, Some(response))) => resultFromEmailOption(response.email)
        case Some(PaymentAPIResponse(_, Some(error), _)) if (error.errorName.contains("PAYMENT_ALREADY_DONE")) => Redirect("/contribute/one-off/thankyou")
        case _ => paypalErrorPage
      } 
```